File: src/python/algorithms/searching/binary_search.py
#!/bin/python


def binary_search(data, target):
    l = 0
    r = len(data) - 1
    while r >= l:
        guess = int((l + r) / 2)
        if data[guess] < target:
            l = guess + 1
        elif data[guess] > target:
            r = guess - 1
        else:
            return guess

    return None

File: src/csharp/Algorithms/Searching/BinarySearch.cs
using System;
using System.Collections.Generic;
using DataStructures;
using Common;


namespace Algorithms.Searching
{
	
	public class BinarySearch : ISearchingAlgorithm
	{

		public int Execute<T>(IDataStructure<T> data, T target)
		{
			int min = 0;
			int max = data.Count;

			while (min <= max) {

				int guess = (min + max) / 2;

				var comparison = Comparer<T>.Default.Compare(data[guess], target);

				if (comparison < 0)
					min = guess + 1;
				else if (comparison > 0)
					max = guess - 1;
				else
					return guess;

			}

			return data.Count;
		}


		public static int Main(string[] args)
		{
			return Tests.Search(args, new BinarySearch());
		}

	}

}

